### ISSUE TITLE
invalid-method-override when overriding float.__round__

### ISSUE THREAD

**Issue by hudcap** [astral-sh/ty#3086]
### Summary

I found [#166](https://github.com/astral-sh/ty/issues/166), but I don't see anything in there that relates to my issue. I don't really understand all the abstract stuff here, so please forgive me if this is a really dumb question.

MWE
[Playground](https://play.ty.dev/73e0b563-f1f5-4a5b-9063-9142198e6014)
```python
from typing import SupportsIndex


class MyFloat(float):
    def __round__(self, ndigits: SupportsIndex, /) -> float:
        return super().__round__(ndigits)
```

`ty` in VSCode is throwing `invalid-method-override` and pointing to `builtins.pyi` lin 788:
```python
    @overload
    def __round__(self, ndigits: SupportsIndex, /) -> float: ...
```

I copied the exact signature, but it's still throwing the error.

### Version

ty server version: 0.0.23 (be6e94a27 2026-03-13)

**Comment by MeGaGiGaGon** [comment]
The issue is that `__round__` on `float` has overloads, but your implementation only supports one of them, You can see that by the `@overload` on the snippet you copied, though you are missing part of it:
```py
    @overload
    def __round__(self, ndigits: None = None, /) -> int:
        """Return the Integral closest to x, rounding half toward even.

        When an argument is passed, work like built-in round(x, ndigits).
        """

    @overload
    def __round__(self, ndigits: SupportsIndex, /) -> float: ...
```
An overload is a way to implement more specific behavior than you can normally get in a function signature. In the case of `float.__round__`, the overload is used to say "if you don't pass in `ndigits`, you always get an `int`". To read more about overloads see this page: https://docs.python.org/3/library/typing.html#typing.overload

With that in mind, here is what a conforming override would look like:
https://play.ty.dev/6a99ebd2-c56d-4611-937c-a444053daad5
```py
from typing import SupportsIndex, overload

class MyFloat(float):
    @overload
    def __round__(self, ndigits: None = None, /) -> int: ...
    @overload
    def __round__(self, ndigits: SupportsIndex, /) -> float: ...
    def __round__(self, ndigits: SupportsIndex | None = None, /) -> float | int:
        raise NotImplementedError
```
Notice the two `@overload`s that match what is in the type stubs. In addition, because you are writing actual code and not stubs, you need an implementation version, which is usually made by combining all the overloads with unions, like how `None` on the first and `SupportsIndex` on the second became `SupportsIndex | None`.

Also, you may want/need to change the `float`s in the return types to `MyFloat`/`Self`, depending on your implementation.

As for the subpar diagnostic on this, that is already tracked by #1644

**Comment by hudcap** [comment]
Thank you @MeGaGiGaGon for that great explanation!

I hope it's okay that I humbly ask a follow-up. Feel free to ignore.
You are correct that in my actual code, I'm returning `MyFloat`, so I now have (roughly) this:
```python
from typing import SupportsIndex, overload


class MyFloat(float):
    @overload
    def __round__(self, ndigits: None = None, /) -> int: ...
    @overload
    def __round__(self, ndigits: SupportsIndex, /) -> MyFloat: ...
    def __round__(self, ndigits: SupportsIndex | None = None, /) -> MyFloat:
        return MyFloat(super().__round__(ndigits))
```
which makes `ty` happy, but seems wrong to me, because I always return `MyFloat` and never an `int`. From what I read about the Liskov principle, I am violating it, because with no arguments, it returns `MyFloat` instead of `int`, but the `float` in `MyFloat` is really an `int`.
Do I just ignore this and move on? Or am I doing something fundamentally wrong?

**Comment by carljm** [comment]
@hudcap There is one check that we are missing which should error on your example, which is that we should require you to annotate your implementation method as `-> MyFloat | int`, not just `-> MyFloat`. If you change that annotation, all type checkers will accept this, even though you are correct that it actually breaks Liskov (because the implementation does not return `int` under the conditions that the overloads claim.)

The core issue here is that type checkers do not (and practically cannot, in general) check that your overloads correctly describe the behavior of your implementation. Overloads are effectively a way to make unchecked assertions to the type checker about the function behavior. So it is up to you to ensure that your overload definitions are accurate. This is unfortunate, but there isn't a clear and obvious way to improve the situation.

**Comment by hudcap** [comment]
Thank you so much!
